src/iarc7_motion/state_monitor.py

```diff
@@ -0,0 +1,164 @@
+        
+        # info needed to do sanity checking and state monitoring
+        self._roomba_status_sub = rospy.Subscriber(
+            'roombas', OdometryArray, 
+            self._receive_roomba_status)
+
+        self._obstacle_sub = rospy.Subscriber(
+            '/obstacles', OdometryArray, 
+            self._receive_obstacle_status)
+
+        self._current_velocity_sub = rospy.Subscriber(
+            '/odometry/filtered', Odometry,
+            self._recieve_drone_odometry)
+
+        self._drone_arm_status = rospy.Subscriber(
+            '/fc_status', FlightControllerStatus, 
+            self._receive_arm_status)
+
+        try:
+            # task timeout time to decelerate
+            self._task_timeout_deceleration_time = rospy.Duration(
+                rospy.get_param('~task_timeout_deceleration_time'))
+            # minimum safe height to manuever at
+            self._MIN_MANEUVER_HEIGHT = rospy.get_param('~min_maneuver_height')
+
+        except KeyError as e:
+            rospy.logerr('Could not lookup a parameter for State Monitor')
+            raise
+
+    # checks task transitions before executing it
+    def check_transition(self, task):
+        with self._lock: 
+            passed = False
+
+            if self._waiting_on_takeoff: 
+                passed = isinstance(task, TakeoffTask)
+            elif isinstance(task, LandTask):
+                self._waiting_on_takeoff = True
+                passed = True
+            elif self._waiting_on_recovery:
+                passed = isinstance(task, HeightRecoveryTask)
+            elif self._below_min_manuever_height():
+                passed = isinstance(task, TakeoffTask) or isinstance(task, HeightRecoveryTask)
+            else: 
+                passed = True
+
+            if passed:
+                if isinstance(task, TakeoffTask):
+                    self._waiting_on_takeoff = False
+                elif isinstance(task, BlockRoombaTask) or isinstance(task, HitRoombaTask):
+                    self._waiting_on_recovery = True
+                elif isinstance(task, HeightRecoveryTask):
+                    self._waiting_on_recovery = False
+
+            rospy.loginfo("Waiting on Takeoff: " + str(self._waiting_on_takeoff))
+            rospy.loginfo("Waiting on Recovery: " + str(self._waiting_on_recovery))
+
+            return passed
+
+    # checks to see if drone above min manuever height
+    def _below_min_manuever_height(self):
+        with self._lock:
+            if self._drone_odometry is None:
+                return True
+            else:
+                return (self._drone_odometry.pose.pose.position.z 
+                            < self._MIN_MANEUVER_HEIGHT)
+
+    # fills out the rest of Intermediary State for the task
+    def fill_out_transition(self, state):
+        while ((self._drone_odometry is None or self._roombas is None 
+                or self._arm_status is None) and not rospy.is_shutdown()):
```

Comment:
This needs to say something, not error silently.

Reply:
fixed

Reply:
This also needs to be rate limited

Reply:
fixed, rate limited to 5 milliseconds